File: Queue/Stream First Non-repeating/counting_app.cpp
#include<bits/stdc++.h>
using namespace std;

//Time complexity :- O(N)
//Space complexity :- O(N) 

// Approach :-
// Add all the char which has frequency as 1, traversing from left to right 
// for char check which char is present at front of queue 

//link :- https://www.geeksforgeeks.org/problems/first-non-repeating-character-in-a-stream1216/1

class Solution {
  public:
    string firstNonRepeating(string &s) {
        string res = "";
        queue<char> q;              // store the char having non-repeating freq
        int freq[26] = {0};         // for counting the freq of each element 
        
        for(auto ch: s){
            // increment the freq of current char 
            freq[ch - 'a'] += 1;
            
            // remove all the element from the queue which has repeating freq
            while(!q.empty() && freq[q.front() - 'a'] > 1){
                q.pop();
            }
            
            // push the current element into the queue only if it has single freq
            if(freq[ch - 'a'] == 1)
                q.push(ch);
            
            // add the front element to queue or '#' if it's empty 
            if(!q.empty())          res += q.front();
            else                    res += '#';
        }
        return res;
    }
};

File: Array/Logical or Intutive/Longest Bitonic Subarray/counting_app.java
import java.util.*

// Time complexity :- O(n)
// Space complexity :- O(n)

// Approach :-  
// travese from right to left and store the non-decreasing subarray count in the corresponding index 
// in similar way we will maintain non-increasing count from left to right
// considering current element as a peak of the bitonic array which length equal to length from left + length from right - 1
// maintain maximum length among all the values 


//Link :- https://www.geeksforgeeks.org/problems/maximum-length-bitonic-subarray5730/1


class Solution {
    public int bitonic(int[] arr) {
        int n = arr.length;
        int[] decCnt = new int[n];      // maitain count of decreasing continous subarray length
        
        // traverse from right to left         
        for(int i=n-1; i>=0; --i){
            // check if the current element if on decreasing curve 
            if(i < n-1 && arr[i] >= arr[i+1]){
                decCnt[i] = 1 + decCnt[i+1];
            }else{
                decCnt[i] = 1;      // the default size of subarray will be 1 
            }
        }
        int incCnt = 0;
        int res = 0;
        
        // traverse from left to right 
        for(int i=0; i<n; ++i){
            // check if the current element if on increasing curve 
            if(i > 0 && arr[i] >= arr[i-1]){
                incCnt += 1;
            }else{
                incCnt = 1;     // the default size of subarray will be 1 
            }
            // find the length of current bitonic sequence 
            int len = Math.abs(incCnt + decCnt[i] - 1);
            // maintain max length 
            res = Math.max(res, len);
        }
        return res;
    }
}